ouroboros-consensus/src/unstable-consensus-testlib/Test/CsjModel/Jumping.hs

```diff
@@ -0,0 +1,249 @@
+  where
+    ImmutableJumpInfo immAnchor _immHdrSt = immJumpInfo
+
+    imm = pointToWithOriginRealPoint $ AF.anchorToPoint immAnchor
+
+    stepSTM = stimulateSTM env varCsj varInboxes pid imm
+
+unregisterClient ::
+     (
+       Ord pid, Show pid
+     ,
+       Show (Header blk)
+     ,
+       BlockSupportsProtocol blk, HV.HasAnnTip blk
+     ,
+       IOLike m
+     )
+  => CsjEnv (RealPoint blk)
+  -> StrictTVar m (F (CsjState pid) blk)
+  -> StrictTVar m (Map pid (Inbox m blk))
+     -- ^ CSJ client inboxes
+  -> pid
+  -> WithOrigin (RealPoint blk)
+     -- ^ current immutable tip
+  -> STM m ()
+unregisterClient env varCsj varInboxes pid imm = do
+    stepSTM Disconnect
+    modifyTVar varInboxes $ Map.delete pid
+  where
+    stepSTM = stimulateSTM env varCsj varInboxes pid imm
+
+rotateDynamo ::
+    (
+       Ord pid, Show pid
+     ,
+       Show (Header blk)
+     ,
+       BlockSupportsProtocol blk, HV.HasAnnTip blk
+     ,
+       IOLike m
+     )
+  => CsjEnv (RealPoint blk)
+  -> StrictTVar m (F (CsjState pid) blk)
+  -> StrictTVar m (Map pid (Inbox m blk))
+  -> pid
+  -> WithOrigin (RealPoint blk)
+     -- ^ current immutable tip
+  -> STM m ()
+rotateDynamo env varCsj varInboxes pid imm = do
+    stepSTM Starvation
+  where
+    stepSTM = stimulateSTM env varCsj varInboxes pid imm
+
+csjModelJumping ::
+     (
+       Ord pid, Show pid
+     ,
+       Show (Header blk)
+     ,
+       BlockSupportsProtocol blk, HV.HasAnnTip blk
+     ,
+       IOLike m
+     )
+  => STM m (WithOrigin (RealPoint blk))
+     -- ^ current immutable tip
+  -> CsjEnv (RealPoint blk)
+  -> StrictTVar m (F (CsjState pid) blk)
+  -> StrictTVar m (Map pid (Inbox m blk))
+     -- ^ CSJ client inboxes
+  -> pid
+  -> Jumping m blk
+csjModelJumping getImmTip env varCsj varInboxes pid = Jumping {
+    jgNextInstruction   = atomically $ do
+        inboxes <- readTVar varInboxes
+        let inbox = inboxes Map.! pid
+        reaction <- readTVar inbox >>= \case
+            Strict.Nothing       -> retry
+            Strict.Just reaction -> pure reaction
+        writeTVar inbox Strict.Nothing
+        case reaction of
+            Continue                 -> pure RunNormally
+            Disengage                -> pure RunNormally
+            MsgFindIntersect purp wp -> do
+                stepSTM Offered
+                let f = case purp of
+                        ToBisect  -> JumpTo          
+                        ToPromote -> JumpToGoodPoint
+                pure $ JumpInstruction $ f $ wpPayload wp
+            Stop p                   ->
+                pure $ StopRunningNormally $ withOriginRealPointToPoint p
+  ,
+    jgOnAwaitReply      = step $ ChainSyncReply MsgAwaitReply
+  ,
+    jgOnRecvRollForward =
+        step . ChainSyncReply . MsgRollForwardSTART . castRealPoint
+  ,
+    jgOnRollForward     = \p ji -> atomically $ do
+        -- Suppress this stimulus if CSJ has already sent 'Stop'; this
```

Comment:
Do `jgOnAwaitReply`, `jgOnRecvRollFordward`, and `jgOnRollBackward` also need to check for `Stop`, since the Diffusion Layer thread is calling the handlers as responses come in to the pipelined `MsgRequestNext` messages?